### ISSUE TITLE
Using CameraDevice

### ISSUE THREAD

**Issue by Tim-Kindberg** [pedroSG94/rtmp-rtsp-stream-client-java#574]
Hi,

I'm developing an app in Flutter whose camera plugin uses a android.hardware.camera2.CameraDevice, and I'd like to stick with that. As with a MediaRecorder, a component wanting to consume the camera's output presents a Surface to it. 

I'm making a LiveStreamer class (based on this rtmp/rtsp library) that exports a surface to the CameraDevice, and which retrieves the data from that Surface and passes it directly for RTMPS encoding by a com.pedro.encoder.video.VideoEncoder. No need for filters - although I might add that later. I'm new to OpenGL and I'm having trouble navigating my way around this library to find what I need to do - and to eliminate what I don't need to do. Could someone please point me at classes I need to use or learn from - or to other references on the web?

Many thanks in advance.

**Comment by pedroSG94** [comment]
If you only need video. You need create your own encoder using VideoEncoder class where you should get a surface after start encoder:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/6898d004827628896ceb8e557cb52aa10a09d6f5/encoder/src/main/java/com/pedro/encoder/video/VideoEncoder.java#L247

If you render this surface with your camera plugin you have your h264 buffer that you can send directly to SrsFlvMuxer like here:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/rtmp/RtmpCamera1.java#L167
Remember that before send video data you need set audio and video info to SrsFlvMuxer and start it:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/rtmp/RtmpCamera1.java#L120
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/rtmp/RtmpCamera1.java#L161
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/rtmp/RtmpCamera1.java#L126

**Comment by Tim-Kindberg** [comment]
Thanks very much. I'll give this a go. I do need Audio too but I think I see how to do that.

**Comment by Tim-Kindberg** [comment]
Hi Pedro, I'm getting an exception when trying to build an android CameraCaptureSession with both my preview surface and the one provided by VideoEncoder:

> android.hardware.camera2.CameraAccessException: CAMERA_ERROR (3): submitRequestList - cannot use a surface that wasn't configured

I'm using the existing Flutter camera plugin logic for creating a capture session, the same as for the case when it records to an android MediaRecorder while previewing, except I use the Surface from the VideoEncoder not the MediaRecorder. 

Prior to trying to use the surface, I call VideoEncoder's prepareVideo method with the same default parameters you use in Camera1Base, and a format value of FormatVideoEncoder.SURFACE. I've looked at VideoEncoder, which obtains the surface from its instance of MediaCodec. It doesn't seem to do anything further in the way of configuring it.

Am I missing something? BTW, you said above that 

> you should get a surface after start encoder

but did you mean after preparing the encoder? FWIW, I do start the encoder before handing the surface to the capture session.

**Comment by pedroSG94** [comment]
That surface can be renderer after prepareVideo but only produce frames after start videoEncoder for this reason this isn't relevant if you set it to camera before start encoder but no produce any frames (I do it before start encoder with surfaceview and after start encoder with opengl).
I don't know where is the fail your logic seems good. I do the same but with my own Camera2 implementation:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/7ec103288e2372d461a19b1e54aabbc6510ce39c/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L435
Basically with an encoder previously prepared (calling prepareVideo method in that class). I'm calling videoEncoder.start() before that in  prepareVideo() (prepareCameraManager()) method I set videoEncoder surface to cameraManager and I start that cameraManager.

If you get a Surface not null from getInputSurface and your prepareVideo return true that means that should work because that surface should be renderable with any API according to Android documentation:
https://developer.android.com/reference/android/media/MediaCodec#createInputSurface()

**Comment by Tim-Kindberg** [comment]
Thanks Pedro, I hadn't spotted your Camera2 code. Anyway, I've fixed it. It turned out that the exception raised ("cannot use a surface that wasn't configured") was misleading. Both surfaces were configured. However, I had goofed and one was 720x480 whereas the other was 480x720. It worked as soon as I swapped one of the them. 

I streamed to Facebook Live. But now I have another problem. The RTMPS connection fails every time after 30 seconds:

> onConnectionFailedRtmp Error send packet: Write error: ssl=0x8da145c0: I/O error during system call, Broken pipe

I'm not setting any timers in my code. Any idea what that could be? I tried with the same stream parameters (i.e. the same Facebook session) from your app under the default configuration and the stream remains uninterrupted.

**Comment by pedroSG94** [comment]
Is it working only 30s? 
This is very weird because that is a secure connection error and you haven't problems with default example.
According to this post:
https://stackoverflow.com/questions/38016109/javax-net-ssl-sslexception-write-error-ssl-0x7f70604080-i-o-error-during-syst
Maybe you send to server larges packets. Can you check video packet size in logcat to confirm?
Also if you are using adaptative bitrate feature delete it and use a low bitrate to test.

**Comment by Tim-Kindberg** [comment]
Hi Pedro, 

In case it's relevant: I'm not sending any audio at this stage, just video. I'm working on audio integration currently (from an external USB device). I'm not knowingly setting a variable bitrate - where / how might I be doing that inadvertently? I do see messages about new bitrates sent to the ConnectCheckerRtmp and there's an instance of that in the logcat output below.

I changed the rate I was using of 6Mbit/sec down to 2Mbit/sec and it still fails after 30secs like it always does. Here's the last bit of the logcat output before the fail. That "onFCUnpublish" looks interesting:

> 
> 2020-05-28 09:36:09.214 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 10188
> 2020-05-28 09:36:09.246 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 5432
> 2020-05-28 09:36:09.279 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 5791
> 2020-05-28 09:36:09.317 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 13164
> 2020-05-28 09:36:09.344 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 12144
> 2020-05-28 09:36:09.346 21830-24715/com.foo.bar W/RtmpConnection: handleRxPacketLoop(): Not handling unimplemented/unknown packet of type: ACKNOWLEDGEMENT
> 2020-05-28 09:36:09.368 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 5680
> 2020-05-28 09:36:09.405 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 7078
> 2020-05-28 09:36:09.448 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 8329
> 2020-05-28 09:36:09.477 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 3635
> 2020-05-28 09:36:09.505 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 8331
> 2020-05-28 09:36:09.538 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 13658
> 2020-05-28 09:36:09.569 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 14965
> 2020-05-28 09:36:09.604 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 13188
> 2020-05-28 09:36:09.646 21830-24681/com.foo.bar I/SrsFlvMuxer: worker: send frame type=9, dts=30314, size=131072B
> 2020-05-28 09:36:09.648 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 30324
> 2020-05-28 09:36:09.649 21830-24681/com.foo.bar I/M2mLiveStreamer: onNewBitrateRtmp 2312336
> 2020-05-28 09:36:09.682 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 8088
> 2020-05-28 09:36:09.710 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 4213
> 2020-05-28 09:36:09.736 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 6838
> 2020-05-28 09:36:09.779 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 8072
> 2020-05-28 09:36:09.813 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 8736
> 2020-05-28 09:36:09.839 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 5884
> 2020-05-28 09:36:09.865 21830-24715/com.foo.bar E/RtmpConnection: handleRxInvoke(): Unknown/unhandled server invoke: RTMP Command (command: onFCUnpublish, transaction ID: 0)
> 2020-05-28 09:36:09.866 21830-24715/com.foo.bar I/RtmpConnection: handleRxInvoke: Got result for invoked method: publish
> 2020-05-28 09:36:09.866 21830-24715/com.foo.bar W/RtmpConnection: handleRxInvoke(): '_result' message received for unknown method: publish
> 2020-05-28 09:36:09.867 21830-24715/com.foo.bar D/RtmpConnection: handleRxInvoke(): onStatus NetStream.Unpublish.InternalKill
> 2020-05-28 09:36:09.871 21830-24681/com.foo.bar D/RtmpConnection: wrote packet: RTMP Video, size: 8091
> 2020-05-28 09:36:09.912 21830-24681/com.foo.bar I/M2mLiveStreamer: onConnectionFailedRtmp Error send packet: Write error: ssl=0x8ea71380: I/O error during system call, Broken pipe

FWIW, I changed to a really low rate: 500,000 bits/sec. Still fails.

**Comment by pedroSG94** [comment]
As you said, in your logs I can see that Facebook send you a request to stop stream "onFCUnpublish".

I'm 99% sure that Facebook stream need audio and video to work (at least YouTube and Twitch need it and this platform is similar) so when you only send video but audio is not received, Facebook server send you a request to stop stream. But my library doesn't implement this method, so after that, Facebook forcibly stop the connection.

I recommend you finish your audio implementation but if you need so much time maybe you can send device audio like I'm doing in my library to discard this error possibility. 
Basically you only need create a MicrophoneManager and send that data to an AudioEncoder that work similar that VideoEncoder but you need set audio frames from MicrophoneManager instead of render a surface.

**Comment by Tim-Kindberg** [comment]
Thanks Pedro, I did wonder about that but couldn't think why audio would be mandatory. I'll try again after adding audio. Your help and prompt replies are very much appreciated.

**Comment by pedroSG94** [comment]
I think that if you are using video, audio is mandatory, but only audio is possible because in 2017 seems that podcast stream was added.
Other way to test it could be never send audio frames with my app example to discard it (you only need comment this line):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/rtmp/RtmpCamera1.java#L157
Anyway you can ask in Facebook developer forum to be 100% sure:
https://developers.facebook.com/community/?locale=es_ES

**Comment by Tim-Kindberg** [comment]
Hi Pedro, thanks. I can confirm that the Facebook Live stream doesn't get closed when I send audio as well.
